File: testdata/samplecdp/config/evolution/submissions/accepted/Evolution.java

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.StringTokenizer;

/**
 * World Finals 2015 problem solution: PARALLEL EVOLUTION
 * 
 * @author Martin Kacer
 */
public class Evolution {
	StringTokenizer st = new StringTokenizer("");
	BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
	public static void main(String[] args) throws Exception {
		Evolution instance = new Evolution();
		while (instance.run()) { /*repeat*/ }
	}
	String nextToken() throws Exception {
		while (!st.hasMoreTokens()) {
			String ln = input.readLine();
			if (ln == null) return null;
			st = new StringTokenizer(ln);
		}
		return st.nextToken();
	}
	int nextInt() throws Exception {
		String s = nextToken();
		return (s == null) ? -1 : Integer.parseInt(s);
	}
	
	static final char[] EMPTY_SEQ = new char[0];
	
	char[][] samples;
	char[] current;
	int[] parent;

	/** Solve one instance. */
	boolean run() throws Exception {
		int n = nextInt(); if (n < 0) return false;
		samples = new char[n][];
		parent = new int[n];
		current = nextToken().toCharArray();
		for (int i = 0; i < n; ++i)
			samples[i] = nextToken().toCharArray();
		if (!solve())
			System.out.println("Impossible");
		return true;
	}
	
	boolean solve() {
		for (int i = 0; i < samples.length; ++i)
			if (!isSubString(samples[i], current))
				return false;
		Arrays.sort(samples, new LengthComparator());
		char[] s1 = EMPTY_SEQ, s2 = EMPTY_SEQ, common = null;
		int i1 = -1, i2 = -1, ic = -1, ic0 = -1;
		for (int i = 0; i < samples.length; ++i) {
			char[] s = samples[i];
			if (common == null) {
				if (isSubString(s1, s)) {
					if (isSubString(s2, s)) {
						common = s; ic = ic0 = i; // parent[ic0] will be set later
					} else {
						s1 = s; parent[i] = i1; i1 = i;
					}
				} else if (isSubString(s2, s)) {
					s2 = s; parent[i] = i2; i2 = i;
				} else
					return false;
			} else {
				if (isSubString(common, s)) {
					common = s; parent[i] = ic; ic = i;
				} else {
					if (isSubString(s1, s)) {
						s1 = s; parent[i] = i1; i1 = i;
						s2 = common; parent[ic0] = i2; i2 = ic;
					} else if (isSubString(s2, s)) {
						s2 = s; parent[i] = i2; i2 = i;
						s1 = common; parent[ic0] = i1; i1 = ic;
					} else
						return false;
					common = null;
				}
			}
		}
		if (common != null) {
			parent[ic0] = i1; i1 = ic;
		}
		int cnt1 = 0;
		for (int i = i1; i >= 0; i = parent[i]) ++cnt1;
		System.out.println(cnt1 + " " + (samples.length - cnt1));
		List<String> lst = new ArrayList<String>(samples.length);
		for (int i = i1; i >= 0; i = parent[i])
			lst.add(new String(samples[i]));
		Collections.reverse(lst);
		for (String s : lst) System.out.println(s);
		lst.clear();
		for (int i = i2; i >= 0; i = parent[i])
			lst.add(new String(samples[i]));
		Collections.reverse(lst);
		for (String s : lst) System.out.println(s);
		return true;
	}
	
	static boolean isSubString(char[] sub, char[] big) {
		if (sub.length > big.length) return false;
		int i = 0;
		for (int j = 0; i < sub.length && j < big.length; ++j) {
			if (sub[i] == big[j]) ++i;
		}
		return (i == sub.length);
	}

	static class LengthComparator implements Comparator<char[]> {
		public int compare(char[] s1, char[] s2) {
			return s1.length - s2.length;
		}
	}
}

File: testdata/samplecdp/config/evolution/submissions/accepted/evolution.c

/**
 * World Finals 2015 problem solution: PARALLEL EVOLUTION
 * 
 * @author Martin Kacer
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define MAX_SAMPLES 5000
#define MAX_LENGTH 5000

int cnt;
char current[MAX_LENGTH+1];
char samples[MAX_SAMPLES+2][MAX_LENGTH+1];
int len[MAX_SAMPLES], idx[MAX_SAMPLES], next[MAX_SAMPLES+1];

int contains(const char* sup, const char* sub) {
	while (*sup && *sub) {
		if (*sup == *sub) ++sub;
		++sup;
	}
	return !*sub;
}

int length_comparator(const void* a, const void* b) {
	return len[*((const int*)a)] - len[*((const int*)b)];
}

int solve() {
	int i, i1 = MAX_SAMPLES, i2 = MAX_SAMPLES+1, ic = -1, icc = -1;
	for (i = 0; i < cnt; ++i) {
		if (!contains(current, samples[i])) return 0;
		len[i] = strlen(samples[i]);
		idx[i] = i;
	}
	qsort(idx, cnt, sizeof(int), &length_comparator);
	samples[MAX_SAMPLES][0] = samples[MAX_SAMPLES+1][0] = '\0';
	for (i = 0; i < cnt; ++i) {
		if (ic >= 0 ) {
			if (contains(samples[idx[i]], samples[ic])) {
				next[ic] = idx[i]; ic = idx[i];
			} else if (contains(samples[idx[i]], samples[i1])) {
				next[i1] = idx[i]; i1 = idx[i];
				next[i2] = icc; i2 = ic;
				ic = icc = -1;
			} else if (contains(samples[idx[i]], samples[i2])) {
				next[i2] = idx[i]; i2 = idx[i];
				next[i1] = icc; i1 = ic;
				ic = icc = -1;
			} else return 0;
		} else {
			if (contains(samples[idx[i]], samples[i1])) {
				if (contains(samples[idx[i]], samples[i2])) {
					ic = icc = idx[i];
				} else {
				next[i1] = idx[i]; i1 = idx[i];
				}
			} else if (contains(samples[idx[i]], samples[i2])) {
				next[i2] = idx[i]; i2 = idx[i];
			} else return 0;
		}
	}
	if (ic >= 0) {
		next[i2] = icc; i2 = ic;
	}
	ic = 0;
	for (i = MAX_SAMPLES; i != i1; i = next[i]) ++ic;
	printf("%d %d\n", ic, cnt-ic);
	for (i = MAX_SAMPLES; i != i1; i = next[i]) printf("%s\n", samples[next[i]]);
	for (i = MAX_SAMPLES+1; i != i2; i = next[i]) printf("%s\n", samples[next[i]]);
	return 1;
}

int main() {
	int i;
	scanf("%d", &cnt);
	scanf("%s", current);
	for (i = 0; i < cnt; ++i) scanf("%s", samples[i]);
	if (!solve()) printf("Impossible\n");
	return 0;
}
